README.md

```diff
@@ -16,85 +16,96 @@ Without needing to worry about instruction selection, register pressure, and so
+A pipeline is a directed acyclic graph, where the nodes are `buffer_expr`s, and the edges are `func`s.
+A `func` is said to "consume" its input buffer(s), and "produce" its output buffers, of which there can be more than one.
+A node that is not produced by any `func` is a pipeline input, and a node that is not consumed by a `func` is a pipeline output.
 
-The actual implementation of a `func` is a callback taking a single argument `eval_context`.
-This object contains the state of the program at the time of the call.
-Values of any symbol currently in scope at the time of the call can be accessed in the `eval_context`. 
+A function encapsulates:
+
+- a symbolic description of data dependecies
+- a specification of the compute schedule (loop blocking)
+- a callback representing the actual implementation.
 
 ### Elementwise example
 Here is an example of a simple pipeline of two 1D elementwise `func`s:
 
 ```c++
 node_context ctx;
 
-auto in = buffer_expr::make(ctx, "in", sizeof(int), 1);
-auto out = buffer_expr::make(ctx, "out", sizeof(int), 1);
-auto intm = buffer_expr::make(ctx, "intm", sizeof(int), 1);
+// input buffer expression
+auto in = buffer_expr::make(ctx, "in", 1, sizeof(int));
+// output buffer expression
+auto out = buffer_expr::make(ctx, "out", 1, sizeof(int));
+// temporary buffer expression
+auto tmp = buffer_expr::make(ctx, "tmp", 1, sizeof(int));
 
-var x(ctx, "x");
+var i(ctx, "i");
+// `mul` is automatically registered as the producer of `tmp`.
+func mul = func::make(multiply_2<int>, {{in, {point(i)}}}, {{tmp, {i}}});
 
-func mul = func::make(multiply_2, {in, {point(x)}}, {intm, {x}});
-func add = func::make(add_1, {intm, {point(x)}}, {out, {x}});
+// one could reuse 'i', but for clarity let's use a new variable.
+var j(ctx, "j");
+// `add` is automatically registered as the producer of `out`.
+func add = func::make(add_1<int>, {{tmp, {point(j)}}}, {{out, {j}}});
 
 pipeline p = build_pipeline(ctx, {in}, {out});
 ```
 
-- `in` and `out` are the input and output buffers.
-- `intm` is the intermediate buffer between the two operations.
-- To describe this pipeline, we need one variable `x`.
-- Both `func` objects have the same signature:
-	- Consume a buffer of `const int`, produce a buffer of `int`.
-	- The output dimension is indexed by `x`, and both operations require a the single point interval `[x, x]` of their inputs.
-	- `multiply_2` and `add_1` are functions implementing this operation.
+Without any further specifications this pipeline would use the following strategy:
 
-The possible implementations of this pipeline vary between two extremes:
-1. Allocating `intm` to have the same size as `out`, and executing all of `mul`, followed by all of `add`.
-2. Allocating `intm` to have a single element, and executing `mul` followed by `add` in a single loop over the output elements.
+- [1]  Materialize the whole logical `tmp` buffer of the same size as `out`, execute all of `mul`, followed by all of `add`.
```

Comment:
This "- [1] " bullet number looks weird, maybe write this with numbered lines:

1. Materialize the whole logical `tmp` buffer of the same size as `out`, execute all of `mul`
2. Execute all of `add`